packages/components/src/number-control/index.tsx

```diff
@@ -178,14 +186,27 @@ function UnforwardedNumberControl(
 			return nextState;
 		};
 
+	const buildSpinHandler =
+		( direction: 'up' | 'down' ) =>
+		( event: ChangeEvent< HTMLInputElement > ) => {
+			let nextValue = isValueEmpty( valueProp ) ? baseValue : valueProp;
+			if ( direction === 'up' ) {
+				nextValue = add( nextValue, baseStep );
+			} else if ( direction === 'down' ) {
+				nextValue = subtract( nextValue, baseStep );
+			}
+			nextValue = constrainValue( nextValue );
+			onChange( String( nextValue ), { event } );
+		};
```

Comment:
I did try a few different approaches here but couldn't figure out anything that felt right.

- We can't dispatch reducer actions from `NumberControl` as the `dispatch` function in `InputControl` isn't available.
- We can't call the reducer manually in `NumberControl` as the existing reducer state in `InputControl` isn't available.
- I don't think it makes sense for `InputControl` to have the spin button functionality since it is supposed to be a generic input. (The docs say it's intended to be an eventual `TextControl` replacement.)

So I ended up settling on this boring repetitive approach. I can DRY it up a bit further by making the reducer and spin buttons use the same logic. I've done that in e63d7ed.

Let me know if you have any other suggestions.

Reply:
> * I don't think it makes sense for `InputControl` to have the spin button functionality since it is supposed to be a generic input. (The docs say it's intended to be an eventual `TextControl` replacement.)

This is the thing — InputControl already contains a lot of the scaffolding that _seems generic_ but basically only makes sense for number inputs (drag gestures, arrow key handlers). One can argue that spin buttons are just as generic as up/down arrow key handlers are. I guess in its current state, it's really a `<input type="potentially anything, extend me!">` control rather than a strict `<input type="text">` control.

Some other things I'm thinking in relation to this:

- This line in Storybook will error when using the big spin buttons. Is that a problem?

    https://github.com/WordPress/gutenberg/blob/19186e5ea776cb8b149d87723f511af32f10e983/packages/components/src/number-control/stories/index.js#L35

- The `isPressEnterToChange` prop currently does not behave as expected when drag gestures and up/down keys are involved. When we fix this logic at some point, will we also have to fix the big spin buttons separately?

So, on the surface it kind of seems like things will generally be simpler if make the big spin buttons "generic" (wink wink) and move them down into InputControl.

But in any case, I think this PR is DRY enough for the time being, and it wouldn't be hard to extract the buttons if we feel like it in the future. (We should look at that `event.target.validity.valid` thing before merging though)

Reply:
> InputControl already contains a lot of the scaffolding that seems generic but basically only makes sense for number inputs (drag gestures, arrow key handlers).

It does yet it seems wrong to me. Instead of putting any more effort into fitting that mold my feeling is `InputControl` ought to be broken into a hook + component combo. That would allow `NumberControl` to use the hook and get direct access to `dispatch` for more flexible extensibility. It would also allow making `InputControl` more generic. I have an old experimental branch doing that and I'd been thinking to make a fresh start on it after `NumberControl` was converted to TS. Now, I'm not sure when I'll find the time to do so.

Reply:
Not necessarily related to the immediate changes that this PR may need to make in order to be merged relatively soon — but my general feeling is also that `InputControl` could/should be simplified:

- the reducer logic seems complicated and potentially over-engineered, I'd love to see if we can refactor it to be simpler
- a lot of number-related functionality are written in input control, while it would make sense to me if they were part of `NumberControl`
- I remember discussing whether the `isPressEnterToChange` prop is necessary at all

Reply:
I see 😅 thanks for the discussion.

It's easier to move the custom spinners from `NumberControl` to `InputControl` later in a BC way than it is to do the opposite, so I'm personally inclined to stick with putting them in `NumberControl` at least for now.

> This line in Storybook will error when using the big spin buttons. Is that a problem?

Good catch. I'll fix this by overriding `event.target` to be the `<input>`. We do this already for the drag events.

> The isPressEnterToChange prop currently does not behave as expected when drag gestures and up/down keys are involved. When we fix this logic at some point, will we also have to fix the big spin buttons separately?

I would expect that setting `isPressEnterToChange` should only make it so that `onChange` is not called while the user is typing a number and should not have any effect on the spin controls, no?

> the reducer logic seems complicated and potentially over-engineered, I'd love to see if we can refactor it to be simpler

Yeah not a huge fan of this reducer pattern. I don't really see why `InputControl` couldn't be a standard managed component with callback props for each of the actions (`onPressDown`, `onPressUp`, etc.)